dvc/stage_cache.py

```diff
@@ -0,0 +1,72 @@
+import os
+import json
+import hashlib
+
+from dvc.utils.fs import makedirs
+
+
+def _sha256(string):
+    return hashlib.sha256(string.encode()).hexdigest()
+
+
+def _get_hash(stage):
+    if not stage.cmd or not stage.deps or not stage.outs:
+        return None
+
+    string = _sha256(stage.cmd)
+    for dep in stage.deps:
+        if not dep.def_path or not dep.get_checksum():
+            return None
+
+        string += _sha256(dep.def_path)
+        string += _sha256(dep.get_checksum())
+
+    for out in stage.outs:
+        if not out.def_path:
+            return None
+
+        string += _sha256(out.def_path)
+
+    return _sha256(string)
+
+
+def _get_cache(stage):
```

Comment:
@skshetry Would it make sense to use the same format for the lockfile, btw?

Reply:
Also, need to think if any other fields really matter here for reproducibility. So far it seems like the top-level md5 that we current have in our dvcfiles is really only needed for cmd, as nothing else (except for `always_changed` really matters, unless someone is explicitly trying to do something fishy).

Reply:
What about `wdir`?

Reply:
@charlesbaynham `wdir` doesn't seem to matter, as cmd and relpaths we have in dvc-file stay the same.

Reply:
Maybe I'm missing the point, but say cmd is

    cat "../data/file.txt" > "../out.txt"

Even if `file.txt` and `out.txt` are unchanged and `cmd` is unchanged too, I'd get a different result by changing `wdir` (i.e. I'd get an error since the file wouldn't be found anymore).

Reply:
@charlesbaynham dvcfiles contain paths that are relative to wdir, so for your command `dep` would be `../data/file.txt` and out would be `../out.txt`. And no matter where you move the wdir, if dep is at the same relative path, then output will be at the same relative path.

Reply:
Ah I see what you're saying, sorry for the bother